tiatoolbox/annotation/storage.py

```diff
@@ -0,0 +1,1034 @@
+from abc import ABC
+from typing import (
+    Union,
+    Tuple,
+    List,
+    Dict,
+    Iterable,
+    Optional,
+    Any,
+    IO,
+    Callable,
+)
+from numbers import Number
+import pickle
+import io
+import sqlite3
+from itertools import zip_longest
+from pathlib import Path
+import hashlib
+
+
+from shapely import speedups
+from shapely.geometry import Point, Polygon, LineString
+from shapely.geometry import mapping as geometry2feature
+from shapely.geometry import shape as feature2geometry
+
+from shapely import wkt
+import numpy as np
+import pandas as pd
+
+try:
+    import ujson as json  # pyright: reportMissingModuleSource=false
+except ImportError:
+    import json
+import msgpack
+import yaml
+import tables
+
+if speedups.available:
+    speedups.enable()
+
+Geometry = Union[Point, Polygon, LineString]
+BBox = Tuple[Number, Number, Number, Number]
+QueryGeometry = Union[BBox, Geometry]
+
+RESERVED_PROPERTIES = {
+    "class": pd.Int8Dtype(),
+    "class_": pd.Int8Dtype(),
+    "properties": str,
+    "boundary": bytes,
+    "index": int,
+    "x": int,
+    "y": int,
+    "minX": int,
+    "minY": int,
+    "maxX": int,
+    "maxY": int,
+    "type": int,
+    "type_": int,
+}
+
+ASCII_FILE_SEP = "\x1c"
+ASCII_GROUP_SEP = "\x1d"
+ASCII_RECORD_SEP = "\x1e"
+ASCII_UNIT_SEP = "\x1f"
+ASCII_NULL = "\0"
+ISO_8601_DATE_FORMAT = r"%Y-%m-%dT%H:%M:%S.%f%z"
+
+
+class AnnotationStoreABC(ABC):
+    @staticmethod
+    def geometry_hash(geometry: Geometry) -> int:
+        """Create a 64 bit integer hash of a geometry object."""
+        return int.from_bytes(
+            hashlib.md5(geometry.wkb).digest()[:8], "big", signed=True
+        )
+
+    @staticmethod
+    def _int_feature(feature: dict) -> Dict:
+        """Convert feature coordinates to integers."""
+        feature["coordinates"] = np.array(feature["coordinates"]).astype(int).tolist()
+        return feature
+
+    @classmethod
+    def open(cls, io: Union[Path, str, IO]) -> "AnnotationStoreABC":
+        """Load a store object from a path or file-like object."""
+        raise NotImplementedError()
+
+    def __init__(self) -> None:
+        raise NotImplementedError()
+
+    def __len__(self) -> int:
+        """Return the number of annotations in the store."""
+        raise NotImplementedError()
+
+    @staticmethod
+    def _iterfy(obj: Union[Any, Iterable[Any]]) -> Iterable[Any]:
+        """Make passed objects into an iterable.
+
+        Expected behaviour:
+        - Passing None returns a zero length iterable.
+        - Passing an interable returns the passed iterable unchanged.
+        - Passing a non-iterable (e.g. and int) returns a single element iterator.
+        """
```

Comment:
empty line

Reply:
What do you mean by this? There is an empty line at the end of this docstring

Reply:
There should be an empty line at the end of docstring. I may have mistakenly marked it.